### PR TITLE
[HGCAL] SimTracksters from Caloparticles

### PR DESCRIPTION
**Pull request by felicepantaleo**
#### PR description:
This PR improves SimTracksters by:
  - taking into account the correct fraction of the energy of the layer cluster for the calculation of the energy of the trackster and its PCA (data format change was needed)
  - using the caloparticle as a source for layercluster if the caloparticle crossed the CALO volume boundary, otherwise use the simclusters. 
  - I expect changes in the number of simtracksters (I expect a smaller number of them)

#### PR validation:
validated using `23293.0`

### CODE REVIEW

**Comment by felicepantaleo** [comment]
@cmsbuild please test

File: DataFormats/HGCalReco/interface/Trackster.h [hunk 1]
```diff
@@ -153,7 +153,7 @@ namespace ticl {
     // The vertices of the DAG are the indices of the
     // 2d objects in the global collection
     std::vector<unsigned int> vertices_;
-    std::vector<uint8_t> vertex_multiplicity_;
+    std::vector<float> vertex_multiplicity_;
```
Comment (slava77) [inline DataFormats/HGCalReco/interface/Trackster.h:156]: I'm curious if the schema evolution handles this appropriately.
@felicepantaleo please clarify/confirm that tracksters from the old files are still correctly readable.

**Comment by kpedro88** [comment]
@felicepantaleo you need to propagate the premixing modification to the new `caloparticles` parameter: https://github.com/cms-sw/cmssw/blob/5a23cbc97426e3f685f99c724acbcaab84592f17/RecoHGCal/TICL/python/SimTracksters_cff.py#L23

File: RecoHGCal/TICL/plugins/TrackstersFromSimClustersProducer.cc [hunk 2]
```diff
@@ -104,8 +114,11 @@ void TrackstersFromSimClustersProducer::fillDescriptions(edm::ConfigurationDescr
   desc.add<edm::InputTag>("time_layerclusters", edm::InputTag("hgcalLayerClusters", "timeLayerCluster"));
   desc.add<edm::InputTag>("filtered_mask", edm::InputTag("filteredLayerClustersSimTracksters", "ticlSimTracksters"));
   desc.add<edm::InputTag>("simclusters", edm::InputTag("mix", "MergedCaloTruth"));
+  desc.add<edm::InputTag>("caloparticles", edm::InputTag("mix", "MergedCaloTruth"));
```
Comment (lecriste) [inline RecoHGCal/TICL/plugins/TrackstersFromSimClustersProducer.cc:117]: You need something like [this](https://github.com/cms-sw/cmssw/blob/08fd7d924f8ef28001c9d0fec6df076fb368dce5/SimDataFormats/Associations/python/LCToCPAsssociation_cfi.py#L10-L12) to handle the premixing case.

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by felicepantaleo** [comment]
@cmsbuild please test

File: DataFormats/HGCalReco/interface/Trackster.h (hunk 1 above)
Comment (felicepantaleo) [reply DataFormats/HGCalReco/interface/Trackster.h:156]: I could run the trackster validation on old files with this branch.

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by felicepantaleo** [comment]
ok, I think I am done with it...

**Comment by jpata** [comment]
+reconstruction
- no unexpected/relevant changes

**Comment by srimanob** [comment]
Hi @felicepantaleo 

Is this something you expect?
https://tinyurl.com/yf2sc7ln

Screen Shot 2564-05-07 at 17 57 01

**Comment by felicepantaleo** [comment]
@srimanob yes, definitely

**Comment by qliphy** [comment]
+1

**Pull request merged**